File: Crawls_NCBI_ATAC_scRNA_Dev.py
import pandas as pd
import requests
from bs4 import BeautifulSoup

# 構建搜尋詞
search_term = "((ATAC-seq AND RNA-seq) AND development) AND (human OR mouse) AND \"2000/01/01\"[PDAT]:\"2023/02/24\"[PDAT]"

# 構建NCBI搜尋頁面的網址
base_url = "https://www.ncbi.nlm.nih.gov"
search_url = f"{base_url}/gds/?term={search_term}"

# 下載搜尋頁面HTML並解析
response = requests.get(search_url)
soup = BeautifulSoup(response.content, "html.parser")

# 找到包含搜尋結果的表格
table = soup.find("table", {"class": "tbl"})

# 找到表格中的行
rows = table.findAll("tr")[1:]

# 構建空的DataFrame，將匹配的研究添加到其中
df = pd.DataFrame(columns=["Series", "PMID", "Organism", "Cell line"])

# 遍歷表格中的每一行，提取所需的信息
for row in rows:
    cols = row.findAll("td")
    series = cols[0].find("a").text.strip()
    pmid = cols[1].find("a").text.strip()
    organism = cols[4].text.strip()
    cell_line = cols[5].text.strip()
    
    # 將符合條件的研究添加到DataFrame中
    if organism in ["Homo sapiens", "Mus musculus"] and cell_line != "":
        df = df.append({"Series": series, "PMID": pmid, "Organism": organism, "Cell line": cell_line}, ignore_index=True)

# 保存結果為Excel檔案
df.to_excel("result.xlsx", index=False)

File: Crawls_NCBI_ATAC_scRNA_Dev.R
# 載入必要的套件
library("rentrez")
library("tidyverse")
library("httr")
library("rvest")
library("openxlsx")

# 設定搜尋的關鍵字，時間範圍及查詢的資料庫
query <- "ATAC-seq AND RNA-seq AND development"
mindate <- "2012/01/01"
maxdate <- Sys.Date()
database <- "gds"

# 使用rentrez套件進行查詢
search <- entrez_search(db = database, term = query, mindate = mindate, maxdate = maxdate)
count <- search$count

# 擷取符合條件的GEO編號(GSE ID)
id_list <- entrez_fetch(db = database, web_history = search$web_history, rettype = "xml") %>%
  xml_find_all(".//Id") %>%
  xml_text()

# 設定輸出的Excel檔案名稱和檔案路徑
filename <- "result.xlsx"
path <- "~/"

# 建立一個空的data frame
df <- data.frame(GSE = character(),
                 PMID = character(),
                 Organism = character(),
                 Cell_Line = character(),
                 stringsAsFactors = FALSE)

# 使用for迴圈，逐一爬取每個符合條件的GEO編號(GSE ID)的詳細資訊
for(i in id_list){
  # 用entrez_summary取得每個GEO ID的摘要資訊
  summary <- entrez_summary(db = database, id = i, rettype = "xml")
  # 用rvest套件解析XML檔案中的網頁資訊，並且取得PMC ID
  pmc_id <- read_xml(summary) %>% html_nodes(".//Dbtag[Db='PMC']/Dbtag/Tag/Object-id") %>% html_text()
  # 設定查詢PMC ID的網址
  url <- paste0("https://www.ncbi.nlm.nih.gov/pmc/articles/PMC", pmc_id)
  # 使用httr套件讀取網頁內容
  html <- content(GET(url))
  # 使用rvest套件解析網頁資訊
  organism <- html %>% html_nodes(".content .info-box tr:nth-child(1) td:nth-child(2)") %>% html_text()
  cell_line <- html %>% html_nodes(".content .info-box tr:nth-child(3) td:nth-child(2)") %>% html_text()
  # 將所需資訊儲存到data frame中
  df <- df %>% add_row(GSE = i, PMID = pmc_id, Organism = organism, Cell_Line = cell_line)
}

# 將data frame輸出成Excel檔案
write.xlsx(df, file = paste0(path, filename), row.names = FALSE)

# 提示程式執行完成
cat("Done!")
